File: Windows/src/main/kotlin/app/javaquest/ui/UmlDiagramView.kt
package app.javaquest.ui

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.javaquest.core.UmlBox
import app.javaquest.core.UmlBoxKind
import app.javaquest.core.UmlDiagram
import app.javaquest.core.UmlLayout
import app.javaquest.core.UmlMember
import app.javaquest.core.UmlRelation
import app.javaquest.core.UmlRelationKind
import app.javaquest.core.UmlVisibility

/**
 * Zeichnet ein UML-Klassendiagramm: Kästen mit Namen, Attributen und Methoden,
 * dazu die Linien (Dreiecksspitze, Raute, gestrichelt). Die Anordnung kommt aus
 * `UmlLayout` – dieselbe wie in der Apple-Version.
 */
@Composable
fun UmlDiagramView(diagram: UmlDiagram, modifier: Modifier = Modifier) {
    val layout = UmlLayout.compute(diagram)
    Column(
        modifier.fillMaxWidth().clip(RoundedCornerShape(CardRadius)).background(LocalSurfaces.current.field).padding(16.dp),
    ) {
        Box(Modifier.fillMaxWidth().horizontalScroll(rememberScrollState())) {
            Box(Modifier.width(layout.width.dp + 8.dp).height(layout.height.dp + 8.dp)) {
                Canvas(Modifier.fillMaxSize()) {
                    diagram.relations.forEach { relation ->
                        val from = layout.placed(relation.from)
                        val to = layout.placed(relation.to)
                        if (from != null && to != null) drawRelation(relation, from, to)
                    }
                }
                layout.boxes.forEach { placed ->
                    UmlBoxView(
                        placed.box,
                        Modifier
                            .offset(placed.x.dp, placed.y.dp)
                            .width(placed.width.dp)
                            .height(placed.height.dp),
                    )
                }
            }
        }

        if (diagram.relations.isNotEmpty()) {
            Spacer(Modifier.height(14.dp))
            Eyebrow("Linien im Diagramm", secondaryText)
            Spacer(Modifier.height(6.dp))
            diagram.relations.forEach { relation ->
                RelationLegendRow(relation)
                Spacer(Modifier.height(8.dp))
            }
        }
    }
}

@Composable
private fun UmlBoxView(box: UmlBox, modifier: Modifier) {
    val shape = RoundedCornerShape(8.dp)
    Column(
        modifier
            .clip(shape)
            .background(LocalSurfaces.current.card)
            .border(1.5.dp, Palette.indigo.copy(alpha = 0.45f), shape),
    ) {
        Column(
            Modifier.fillMaxWidth().height(UmlLayout.HEADER_HEIGHT.dp).background(Palette.indigo.copy(alpha = 0.12f)),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            box.kind.stereotype?.let {
                Text(it, fontSize = 11.sp, color = Palette.indigo, fontWeight = FontWeight.Medium)
            }
            Text(
                box.name,
                fontSize = 14.sp,
                fontWeight = FontWeight.Bold,
                fontFamily = FontFamily.Monospace,
                fontStyle = if (box.kind == UmlBoxKind.ABSTRACT) FontStyle.Italic else FontStyle.Normal,
            )
        }
        DividerLine()
        MemberSection(box.fields)
        DividerLine()
        MemberSection(box.methods)
    }
}

@Composable
private fun DividerLine() {
    Box(Modifier.fillMaxWidth().height(1.dp).background(LocalSurfaces.current.divider))
}

@Composable
private fun MemberSection(members: List<UmlMember>) {
    Column(Modifier.fillMaxWidth().padding(horizontal = 10.dp, vertical = UmlLayout.SECTION_PADDING.dp)) {
        if (members.isEmpty()) {
            Spacer(Modifier.height(UmlLayout.ROW_HEIGHT.dp))
        } else {
            members.forEach { member ->
                Row(Modifier.height(UmlLayout.ROW_HEIGHT.dp), verticalAlignment = Alignment.CenterVertically) {
                    Text(
                        member.visibility.symbol,
                        fontSize = 12.sp,
                        fontWeight = FontWeight.Bold,
                        fontFamily = FontFamily.Monospace,
                        color = visibilityColor(member.visibility),
                    )
                    Spacer(Modifier.width(4.dp))
                    Text(
                        member.type?.let { "${member.name}: $it" } ?: member.name,
                        fontSize = 12.sp,
                        fontFamily = FontFamily.Monospace,
                        maxLines = 1,
                    )
                }
            }
        }
    }
}

private fun visibilityColor(visibility: UmlVisibility) = when (visibility) {
    UmlVisibility.PUBLIC -> Palette.success
    UmlVisibility.PRIVATE -> Palette.ember
    UmlVisibility.PROTECTED -> Palette.orange
}

/** Erklärt jede Linie in Alltagssprache – kein Symbol bleibt unerklärt. */
@Composable
private fun RelationLegendRow(relation: UmlRelation) {
    Row(verticalAlignment = Alignment.Top) {
        Canvas(Modifier.size(34.dp, 16.dp).padding(top = 4.dp)) { drawRelationGlyph(relation.kind) }
        Spacer(Modifier.width(8.dp))
        Column {
            Text(
                "${relation.from} → ${relation.to}: ${relation.kind.title}",
                fontSize = 13.sp,
                fontWeight = FontWeight.SemiBold,
            )
            Text(relation.kind.meaning, fontSize = 13.sp, color = secondaryText, lineHeight = 18.sp)
            relation.multiplicity?.let {
                Text(
                    "Vielfachheit $it: ${UmlRelationKind.meaningOfMultiplicity(it)}",
                    fontSize = 13.sp,
                    color = Palette.indigo,
                )
            }
        }
    }
}

// ---------------------------------------------------------------- Zeichnen

private fun DrawScope.drawRelation(relation: UmlRelation, from: UmlLayout.Placed, to: UmlLayout.Placed) {
    val scale = density
    val sameRow = kotlin.math.abs(from.y - to.y) < 1
    val start = if (sameRow) {
        Offset(((from.x + from.width) * scale).toFloat(), ((from.y + from.height / 2) * scale).toFloat())
    } else {
        Offset((from.centerX * scale).toFloat(), (from.y * scale).toFloat())
    }
    val end = if (sameRow) {
        Offset((to.x * scale).toFloat(), ((to.y + to.height / 2) * scale).toFloat())
    } else {
        Offset((to.centerX * scale).toFloat(), (to.bottom * scale).toFloat())
    }

    val path = Path().apply {
        moveTo(start.x, start.y)
        if (sameRow) {
            lineTo(end.x, end.y)
        } else {
            val midY = (start.y + end.y) / 2
            lineTo(start.x, midY)
            lineTo(end.x, midY)
            lineTo(end.x, end.y)
        }
    }
    val effect = if (relation.kind.isDashed) PathEffect.dashPathEffect(floatArrayOf(6f * scale, 4f * scale)) else null
    drawPath(path, Palette.indigo, style = Stroke(width = 1.8f * scale, pathEffect = effect))

    val upwards = !sameRow
    when (relation.kind) {
        UmlRelationKind.EXTENDS, UmlRelationKind.IMPLEMENTS -> drawTriangleHead(end, upwards, scale)
        UmlRelationKind.AGGREGATION, UmlRelationKind.COMPOSITION ->
            drawDiamondHead(end, upwards, relation.kind == UmlRelationKind.COMPOSITION, scale)
        UmlRelationKind.ASSOCIATION, UmlRelationKind.DEPENDENCY -> drawArrowHead(end, upwards, scale)
    }
}

private fun DrawScope.drawTriangleHead(point: Offset, upwards: Boolean, scale: Float) {
    val size = 11f * scale
    val path = Path().apply {
        moveTo(point.x, point.y)
        if (upwards) {
            lineTo(point.x - size * 0.7f, point.y + size)
            lineTo(point.x + size * 0.7f, point.y + size)
        } else {
            lineTo(point.x - size, point.y - size * 0.7f)
            lineTo(point.x - size, point.y + size * 0.7f)
        }
        close()
    }
    drawPath(path, Color.White)
    drawPath(path, Palette.indigo, style = Stroke(width = 1.8f * scale))
}

private fun DrawScope.drawDiamondHead(point: Offset, upwards: Boolean, filled: Boolean, scale: Float) {
    val length = 16f * scale
    val width = 9f * scale
    val path = Path().apply {
        moveTo(point.x, point.y)
        if (upwards) {
            lineTo(point.x - width / 2, point.y + length / 2)
            lineTo(point.x, point.y + length)
            lineTo(point.x + width / 2, point.y + length / 2)
        } else {
            lineTo(point.x - length / 2, point.y - width / 2)
            lineTo(point.x - length, point.y)
            lineTo(point.x - length / 2, point.y + width / 2)
        }
        close()
    }
    drawPath(path, if (filled) Palette.indigo else Color.White)
    drawPath(path, Palette.indigo, style = Stroke(width = 1.8f * scale))
}

private fun DrawScope.drawArrowHead(point: Offset, upwards: Boolean, scale: Float) {
    val size = 9f * scale
    val path = Path().apply {
        if (upwards) {
            moveTo(point.x - size * 0.6f, point.y + size)
            lineTo(point.x, point.y)
            lineTo(point.x + size * 0.6f, point.y + size)
        } else {
            moveTo(point.x - size, point.y - size * 0.6f)
            lineTo(point.x, point.y)
            lineTo(point.x - size, point.y + size * 0.6f)
        }
    }
    drawPath(path, Palette.indigo, style = Stroke(width = 1.8f * scale))
}

/** Kleines Vorschaubild der Linienart für die Legende. */
private fun DrawScope.drawRelationGlyph(kind: UmlRelationKind) {
    val y = size.height / 2
    val effect = if (kind.isDashed) PathEffect.dashPathEffect(floatArrayOf(4f, 3f)) else null
    drawLine(Palette.indigo, Offset(0f, y), Offset(size.width - 10f, y), strokeWidth = 1.6f, pathEffect = effect)

    when (kind) {
        UmlRelationKind.EXTENDS, UmlRelationKind.IMPLEMENTS -> {
            val head = Path().apply {
                moveTo(size.width, y); lineTo(size.width - 10f, y - 5f); lineTo(size.width - 10f, y + 5f); close()
            }
            drawPath(head, Color.White)
            drawPath(head, Palette.indigo, style = Stroke(width = 1.6f))
        }
        UmlRelationKind.AGGREGATION, UmlRelationKind.COMPOSITION -> {
            val head = Path().apply {
                moveTo(size.width, y); lineTo(size.width - 6f, y - 4f)
                lineTo(size.width - 12f, y); lineTo(size.width - 6f, y + 4f); close()
            }
            drawPath(head, if (kind == UmlRelationKind.COMPOSITION) Palette.indigo else Color.White)
            drawPath(head, Palette.indigo, style = Stroke(width = 1.6f))
        }
        UmlRelationKind.ASSOCIATION, UmlRelationKind.DEPENDENCY -> {
            val head = Path().apply {
                moveTo(size.width - 8f, y - 4f); lineTo(size.width, y); lineTo(size.width - 8f, y + 4f)
            }
            drawPath(head, Palette.indigo, style = Stroke(width = 1.6f))
        }
    }
}

File: JavaQuest/DesignSystem/UMLDiagramView.swift
import SwiftUI
import JavaQuestKit

/// Zeichnet ein UML-Klassendiagramm: Kästen mit Namen, Attributen und Methoden,
/// dazu die Linien zwischen ihnen (Dreiecksspitze, Raute, gestrichelt).
/// Die Anordnung kommt aus `UMLLayout` – auf allen Plattformen dieselbe.
struct UMLDiagramView: View {
    let diagram: UMLDiagram

    private var layout: UMLLayout.Result { UMLLayout.compute(diagram) }

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            ScrollView(.horizontal, showsIndicators: true) {
                ZStack(alignment: .topLeading) {
                    RelationLayer(diagram: diagram, layout: layout)
                    ForEach(layout.boxes, id: \.box.name) { placed in
                        UMLBoxView(box: placed.box)
                            .frame(width: placed.width, height: placed.height)
                            .offset(x: placed.x, y: placed.y)
                    }
                }
                .frame(width: max(layout.width, 1), height: max(layout.height, 1))
                .padding(.vertical, 8)
                .padding(.horizontal, 2)
            }

            if !diagram.relations.isEmpty {
                VStack(alignment: .leading, spacing: 8) {
                    Text("LINIEN IM DIAGRAMM")
                        .font(.caption.weight(.heavy)).tracking(1.1)
                        .foregroundStyle(.secondary)
                    ForEach(diagram.relations) { relation in
                        RelationLegendRow(relation: relation)
                    }
                }
            }
        }
        .padding(14)
        .background(Theme.fieldBackground, in: RoundedRectangle(cornerRadius: Theme.cornerRadius, style: .continuous))
    }
}

/// Ein einzelner Klassenkasten mit seinen drei Fächern.
private struct UMLBoxView: View {
    let box: UMLDiagram.Box

    var body: some View {
        VStack(spacing: 0) {
            VStack(spacing: 1) {
                if let stereotype = box.kind.stereotype {
                    Text(stereotype)
                        .font(.caption2.weight(.medium))
                        .foregroundStyle(Theme.indigo)
                }
                Text(box.name)
                    .font(.system(.subheadline, design: .monospaced).weight(.bold))
                    .italic(box.kind == .abstractType)
            }
            .frame(maxWidth: .infinity)
            .frame(height: UMLLayout.headerHeight)
            .background(Theme.indigo.opacity(0.12))

            Divider()
            section(box.fields)
            Divider()
            section(box.methods)
        }
        .background(Theme.cardBackground)
        .overlay(
            RoundedRectangle(cornerRadius: 8, style: .continuous)
                .strokeBorder(Theme.indigo.opacity(0.45), lineWidth: 1.5)
        )
        .clipShape(RoundedRectangle(cornerRadius: 8, style: .continuous))
    }

    @ViewBuilder
    private func section(_ members: [UMLDiagram.Member]) -> some View {
        VStack(alignment: .leading, spacing: 0) {
            if members.isEmpty {
                Text(" ")
                    .font(.system(.caption, design: .monospaced))
                    .frame(height: UMLLayout.rowHeight)
            } else {
                ForEach(members) { member in
                    HStack(spacing: 4) {
                        Text(member.visibility.rawValue)
                            .font(.system(.caption, design: .monospaced).weight(.bold))
                            .foregroundStyle(color(for: member.visibility))
                        Text(member.type.map { "\(member.name): \($0)" } ?? member.name)
                            .font(.system(.caption, design: .monospaced))
                            .lineLimit(1)
                            .minimumScaleFactor(0.8)
                        Spacer(minLength: 0)
                    }
                    .frame(height: UMLLayout.rowHeight)
                    .accessibilityLabel(Text("\(member.line). \(member.visibility.meaning)"))
                }
            }
        }
        .padding(.horizontal, 10)
        .padding(.vertical, UMLLayout.sectionPadding)
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    private func color(for visibility: UMLDiagram.Visibility) -> Color {
        switch visibility {
        case .publicAccess: Theme.success
        case .privateAccess: Theme.ember
        case .protectedAccess: Theme.orange
        }
    }
}

/// Die Linien zwischen den Kästen, inklusive Pfeilspitzen, Rauten und Beschriftungen.
private struct RelationLayer: View {
    let diagram: UMLDiagram
    let layout: UMLLayout.Result

    var body: some View {
        Canvas { context, _ in
            for relation in diagram.relations {
                guard let from = layout.placed(named: relation.from),
                      let to = layout.placed(named: relation.to) else { continue }
                draw(relation, from: from, to: to, in: &context)
            }
        }
    }

    private func draw(_ relation: UMLDiagram.Relation, from: UMLLayout.Placed, to: UMLLayout.Placed, in context: inout GraphicsContext) {
        // Von der Oberkante des Kindes zur Unterkante des Ziels – oder seitlich, wenn sie nebeneinander liegen.
        let sameRow = abs(from.y - to.y) < 1
        let start = sameRow
            ? CGPoint(x: from.x + from.width, y: from.y + from.height / 2)
            : CGPoint(x: from.centerX, y: from.y)
        let end = sameRow
            ? CGPoint(x: to.x, y: to.y + to.height / 2)
            : CGPoint(x: to.centerX, y: to.bottom)

        var path = Path()
        path.move(to: start)
        if sameRow {
            path.addLine(to: end)
        } else {
            let midY = (start.y + end.y) / 2
            path.addLine(to: CGPoint(x: start.x, y: midY))
            path.addLine(to: CGPoint(x: end.x, y: midY))
            path.addLine(to: end)
        }

        let style = StrokeStyle(lineWidth: 1.8, dash: relation.kind.isDashed ? [6, 4] : [])
        context.stroke(path, with: .color(Theme.umlLine), style: style)

        let direction = sameRow ? CGPoint(x: 1, y: 0) : CGPoint(x: 0, y: -1)
        switch relation.kind {
        case .extendsRelation, .implementsRelation:
            drawTriangle(at: end, direction: direction, in: &context)
        case .aggregation, .composition:
            // Die Raute steht am Ganzen – also am Ziel der Linie.
            drawDiamond(at: end, direction: direction, filled: relation.kind == .composition, in: &context)
        case .association, .dependency:
            drawArrow(at: end, direction: direction, in: &context)
        }

        if let text = [relation.label, relation.multiplicity].compactMap({ $0 }).first {
            let midPoint = CGPoint(x: (start.x + end.x) / 2 + 8, y: (start.y + end.y) / 2)
            context.draw(
                Text(text).font(.caption2).foregroundStyle(.secondary),
                at: midPoint,
                anchor: .leading
            )
        }
    }

    private func drawTriangle(at point: CGPoint, direction: CGPoint, in context: inout GraphicsContext) {
        let size: CGFloat = 11
        var path = Path()
        if direction.y != 0 {
            path.move(to: CGPoint(x: point.x, y: point.y))
            path.addLine(to: CGPoint(x: point.x - size * 0.7, y: point.y + size))
            path.addLine(to: CGPoint(x: point.x + size * 0.7, y: point.y + size))
        } else {
            path.move(to: CGPoint(x: point.x, y: point.y))
            path.addLine(to: CGPoint(x: point.x - size, y: point.y - size * 0.7))
            path.addLine(to: CGPoint(x: point.x - size, y: point.y + size * 0.7))
        }
        path.closeSubpath()
        context.fill(path, with: .color(Theme.umlFill))
        context.stroke(path, with: .color(Theme.umlLine), lineWidth: 1.8)
    }

    private func drawDiamond(at point: CGPoint, direction: CGPoint, filled: Bool, in context: inout GraphicsContext) {
        let length: CGFloat = 16
        let width: CGFloat = 9
        var path = Path()
        if direction.y != 0 {
            path.move(to: point)
            path.addLine(to: CGPoint(x: point.x - width / 2, y: point.y + length / 2))
            path.addLine(to: CGPoint(x: point.x, y: point.y + length))
            path.addLine(to: CGPoint(x: point.x + width / 2, y: point.y + length / 2))
        } else {
            path.move(to: point)
            path.addLine(to: CGPoint(x: point.x - length / 2, y: point.y - width / 2))
            path.addLine(to: CGPoint(x: point.x - length, y: point.y))
            path.addLine(to: CGPoint(x: point.x - length / 2, y: point.y + width / 2))
        }
        path.closeSubpath()
        context.fill(path, with: .color(filled ? Theme.umlLine : Theme.umlFill))
        context.stroke(path, with: .color(Theme.umlLine), lineWidth: 1.8)
    }

    private func drawArrow(at point: CGPoint, direction: CGPoint, in context: inout GraphicsContext) {
        let size: CGFloat = 9
        var path = Path()
        if direction.y != 0 {
            path.move(to: CGPoint(x: point.x - size * 0.6, y: point.y + size))
            path.addLine(to: point)
            path.addLine(to: CGPoint(x: point.x + size * 0.6, y: point.y + size))
        } else {
            path.move(to: CGPoint(x: point.x - size, y: point.y - size * 0.6))
            path.addLine(to: point)
            path.addLine(to: CGPoint(x: point.x - size, y: point.y + size * 0.6))
        }
        context.stroke(path, with: .color(Theme.umlLine), lineWidth: 1.8)
    }
}

/// Erklärt jede Linie des Diagramms in Alltagssprache – kein Symbol bleibt unerklärt.
private struct RelationLegendRow: View {
    let relation: UMLDiagram.Relation

    var body: some View {
        HStack(alignment: .top, spacing: 8) {
            RelationGlyph(kind: relation.kind)
                .frame(width: 34, height: 16)
                .padding(.top, 2)
            VStack(alignment: .leading, spacing: 2) {
                Text("\(relation.from) → \(relation.to): \(relation.kind.title)")
                    .font(.caption.weight(.semibold))
                Text(relation.kind.meaning)
                    .font(.caption)
                    .foregroundStyle(.secondary)
                    .fixedSize(horizontal: false, vertical: true)
                if let multiplicity = relation.multiplicity {
                    Text("Vielfachheit \(multiplicity): \(RelationGlyph.meaning(ofMultiplicity: multiplicity))")
                        .font(.caption)
                        .foregroundStyle(Theme.indigo)
                }
            }
        }
    }
}

/// Kleines Vorschaubild der Linienart für die Legende.
private struct RelationGlyph: View {
    let kind: UMLDiagram.RelationKind

    static func meaning(ofMultiplicity value: String) -> String {
        switch value {
        case "1": "genau eins"
        case "0..1": "keins oder eins"
        case "1..*": "mindestens eins"
        case "*": "beliebig viele"
        default: "so viele wie angegeben"
        }
    }

    var body: some View {
        Canvas { context, size in
            let y = size.height / 2
            var line = Path()
            line.move(to: CGPoint(x: 0, y: y))
            line.addLine(to: CGPoint(x: size.width - 10, y: y))
            context.stroke(
                line,
                with: .color(Theme.umlLine),
                style: StrokeStyle(lineWidth: 1.6, dash: kind.isDashed ? [4, 3] : [])
            )

            var head = Path()
            switch kind {
            case .extendsRelation, .implementsRelation:
                head.move(to: CGPoint(x: size.width, y: y))
                head.addLine(to: CGPoint(x: size.width - 10, y: y - 5))
                head.addLine(to: CGPoint(x: size.width - 10, y: y + 5))
                head.closeSubpath()
                context.fill(head, with: .color(Theme.umlFill))
                context.stroke(head, with: .color(Theme.umlLine), lineWidth: 1.6)
            case .aggregation, .composition:
                head.move(to: CGPoint(x: size.width, y: y))
                head.addLine(to: CGPoint(x: size.width - 6, y: y - 4))
                head.addLine(to: CGPoint(x: size.width - 12, y: y))
                head.addLine(to: CGPoint(x: size.width - 6, y: y + 4))
                head.closeSubpath()
                context.fill(head, with: .color(kind == .composition ? Theme.umlLine : Theme.umlFill))
                context.stroke(head, with: .color(Theme.umlLine), lineWidth: 1.6)
            case .association, .dependency:
                head.move(to: CGPoint(x: size.width - 8, y: y - 4))
                head.addLine(to: CGPoint(x: size.width, y: y))
                head.addLine(to: CGPoint(x: size.width - 8, y: y + 4))
                context.stroke(head, with: .color(Theme.umlLine), lineWidth: 1.6)
            }
        }
    }
}

#if DEBUG
#Preview("UML-Diagramm") {
    UMLDiagramView(diagram: UMLDiagram(
        classes: [
            UMLDiagram.Box(name: "Tier", kind: .abstractType,
                           fields: [.init(visibility: .protectedAccess, name: "name", type: "String")],
                           methods: [.init(visibility: .publicAccess, name: "laut()", type: "String")]),
            UMLDiagram.Box(name: "Hund",
                           methods: [.init(visibility: .publicAccess, name: "laut()", type: "String")]),
        ],
        relations: [.init(from: "Hund", to: "Tier", kind: .extendsRelation)]
    ))
    .padding()
}
#endif
